### ISSUE TITLE
Low Precision Inference

### ISSUE THREAD

**Issue by wittmarf** [LeelaChessZero/lc0#794]
Hi there!

Have you considered using less than 16 bits for inference? 

In my experience, it is possible to quantize weights and activation values to 8 bits (and even below) with only a small loss of accuracy.

The Pascal and particularly the Turing GPU is designed to process 8 bit calculations much faster than those with 16 or 32 bits. Another positive effect is less memory usage for the weights.

**Comment by oscardssmith** [comment]
Quantization has been tried. At same number of blocks, it was significantly weaker. My guess is that we would need specific int8.

**Comment by mooskagh** [comment]
Quantization of trained net to 8 bits didn't work indeed.
The agreement now is that the network should have 8-bit precision during training (like a0 did), and requantization of existing 32/16 net is bad.

The experiments were done only once around July 2018, if someone (@wittmarf ?) is in the mood to check again, please do.

**Comment by Ttl** [comment]
I tried int8 quantization in Go last year and got it working so that it was stronger than single precision with equal time and only a little bit weaker with equal nodes.

I didn't manage to get the network to have good enough accuracy just by calibrating the activations and I had to retrain the network with quantization to recover the strength. The drawback is that the network retrained with quantization is now weaker when evaluated with single or half precision.

One very useful trick is to have heads computed with floating point, it's not much slower but improves the accuracy significantly. Input convolution could also be possibly calculated with floating point to improve accuracy.

My branch is in https://github.com/Ttl/leela-zero/tree/cudnn2 if someone is interested in trying to port it to lc0.
